### PR TITLE
[MLIR] EmitC: Add subscript operator

### PR DESCRIPTION
**Pull request by mgehre-amd**
Introduces a SubscriptOp that allows to write IR like
```
func.func @load_store(%arg0: !emitc.array<4x8xf32>, %arg1: !emitc.array<3x5xf32>, %arg2: index, %arg3: index) {
  %0 = emitc.subscript %arg0[%arg2, %arg3] : <4x8xf32>, index, index
  %1 = emitc.subscript %arg1[%arg2, %arg3] : <3x5xf32>, index, index
  emitc.assign %0 : f32 to %1 : f32
  return
}
```
which gets translated into the C++ code
```
v1[v2][v3] = v0[v1][v2];
```

To make this happen, this
- adds the SubscriptOp
- allows the subscript op as rhs of emitc.assign
- updates the emitter to print SubscriptOps

The emitter prints emitc.subscript in a delayed fashing to allow it being used as lvalue.
I.e. while processing
```
%0 = emitc.subscript %arg0[%arg2, %arg3] : <4x8xf32>, index, index
```
it will not emit any text, but record in the `valueMapper` that the name for `%0` is `v0[v1][v2]`, see `CppEmitter::getSubscriptName`. Only when that result is then used (here in `emitc.assign`), that name is inserted into the text.

### CODE REVIEW

**Comment by marbre** [comment]
Thanks for working on this! To make this more useful, I think the subscript op should also be usable on !emitc.ptr pointing to arrays.

**Comment by simon-camp** [comment]
Thanks for the PR @mgehre-amd. Some early comments before looking into the changes in the emitter.

I can imagine different types to be subscriptable with different constraints, some of which you already check for.

###### array type
- indices should be integer like, we use `IntegerIndexOrOpaqueType`in other places for this
- element type and result type should match
- number of indices should match rank of the array type

###### pointer type
- index should be integer like, we use `IntegerIndexOrOpaqueType`in other places for this
- pointee type and result type should match
- only one index allowed

###### opaque type
- indices can be any type
- result can be any type
- number of indices can be variadic since C++23

**Comment by mgehre-amd** [comment]
> Thanks for the PR @mgehre-amd. Some early comments before looking into the changes in the emitter.
> 
> I can imagine different types to be subscriptable with different constraints, some of which you already check for.
> 
> ###### array type
> * indices should be integer like, we use `IntegerIndexOrOpaqueType`in other places for this
> * element type and result type should match
> * number of indices should match rank of the array type
> 
> ###### pointer type
> * index should be integer like, we use `IntegerIndexOrOpaqueType`in other places for this
> * pointee type and result type should match
> * only one index allowed
> 
> ###### opaque type
> * indices can be any type
> * result can be any type
> * number of indices can be variadic since C++23

Thanks for your comments! Do you think we can split support for SubscriptOp for this additional types into separate PRs?

File: mlir/test/Dialect/EmitC/invalid_ops.mlir [hunk 1]
```diff
@@ -387,3 +387,11 @@ func.func @logical_or_resulterror(%arg0: i32, %arg1: i32) {
   %0 = "emitc.logical_or"(%arg0, %arg1) : (i32, i32) -> i32
   return
 }
+
```
Comment (simon-camp) [inline mlir/test/Dialect/EmitC/invalid_ops.mlir:390]: Can you update the error message in `test_assign_to_non_variable` above and add a similar ` test_assign_to_array`.

File: mlir/test/Target/Cpp/subscript.mlir [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+func.func @load_store(%arg0: !emitc.array<4x8xf32>, %arg1: !emitc.array<3x5xf32>, %arg2: index, %arg3: index) {
+  %0 = emitc.subscript %arg0[%arg2, %arg3] : <4x8xf32>
+  %1 = emitc.subscript %arg1[%arg2, %arg3] : <3x5xf32>
+  emitc.assign %0 : f32 to %1 : f32
+  return
+}
+// CHECK: void load_store(float [[ARR1:[^ ]*]][4][8], float [[ARR2:[^ ]*]][3][5],
+// CHECK-SAME:            size_t [[I:[^ ]*]], size_t [[J:[^ ]*]])
+// CHECK-NEXT: [[ARR2]][[[I]]][[[J]]] = [[ARR1]][[[I]]][[[J]]];
```
Comment (simon-camp) [inline mlir/test/Target/Cpp/subscript.mlir:12]: Can you add theses tests. As this op is special cased, testing a few more paths in the emitter is a good idea I think.

```mlir
emitc.func @func1(%arg0 : f32) {
  emitc.return
}

emitc.func @call_arg(%arg0: !emitc.array<4x8xf32>, %i: index, %j: index,
                     %k: index) {
  %0 = emitc.subscript %arg0[%i, %j] : <4x8xf32>
  %1 = emitc.subscript %arg0[%j, %k] : <4x8xf32>

  emitc.call @func1 (%0) : (f32) -> ()
  emitc.call_opaque "func2" (%1) : (f32) -> ()
  emitc.call_opaque "func3" (%0, %1) { args = [1 : index, 0 : index] } : (f32, f32) -> () 
  emitc.return
}
// CHECK: void call_arg(float [[ARR1:[^ ]*]][4][8], size_t [[I:[^ ]*]],
// CHECK-SAME:          size_t [[J:[^ ]*]], size_t [[K:[^ ]*]])
// CHECK-NEXT: func1([[ARR1]][[[I]]][[[J]]]);
// CHECK-NEXT: func2([[ARR1]][[[J]]][[[K]]]);
// CHECK-NEXT: func3([[ARR1]][[[J]]][[[K]]], [[ARR1]][[[I]]][[[J]]]);
```

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td [hunk 3]
```diff
@@ -1125,4 +1125,36 @@ def EmitC_IfOp : EmitC_Op<"if",
...
+    Example:
+
+    ```mlir
+    %i = index.constant 1
+    %j = index.constant 7
+    %0 = emitc.subscript %arg0[%i][%j] : (!emitc.array<4x8xf32>) -> f32
+    ```
+  }];
+  let arguments = (ins Arg<EmitC_ArrayType, "the reference to load from">:$array,
+                       Variadic<Index>:$indices);
```
Comment (simon-camp) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1146]: Use `IntegerIndexOrOpaqueType` here.

**Review by simon-camp — CHANGES_REQUESTED** [review]
Overall looks good to me.

I don't really like that we are accumulating more and more special cases in the emitter. With `emitc.literal`, `emitc.expression` and `emitc.subscript` we now have three ops for which the text emission is deferred in one way or another. I think we can refactor this in a follow up though.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td [hunk 4]
```diff
@@ -1125,4 +1125,36 @@ def EmitC_IfOp : EmitC_Op<"if",
...
+  let description = [{
+    With the `subscript` operation the subscript operator `[]` can be applied
+    to variables or arguments of array type.
+
+    Example:
+
+    ```mlir
+    %i = index.constant 1
+    %j = index.constant 7
+    %0 = emitc.subscript %arg0[%i][%j] : (!emitc.array<4x8xf32>) -> f32
```
Comment (simon-camp) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:1142]:
```suggestion
    %0 = emitc.subscript %arg0[%i, %j] : (!emitc.array<4x8xf32>) -> f32
```

File: mlir/test/Dialect/EmitC/invalid_ops.mlir (hunk 1 above)
Comment (mgehre-amd) [reply mlir/test/Dialect/EmitC/invalid_ops.mlir:390]: I updated the error message in `test_assign_to_non_variable`.
What is the change you'd like to see for
```
func.func @test_assign_to_array(%arg1: !emitc.array<4xi32>) {
  %v = "emitc.variable"() <{value = #emitc.opaque<"">}> : () -> !emitc.array<4xi32>
  // expected-error @+1 {{'emitc.assign' op cannot assign to array type}}
  emitc.assign %arg1 : !emitc.array<4xi32> to %v : !emitc.array<4xi32>
  return
}
```
?

Comment (simon-camp) [reply mlir/test/Dialect/EmitC/invalid_ops.mlir:390]: Oh sorry, I missed that one.

**Review by simon-camp — APPROVED** [review]

**Review by simon-camp — CHANGES_REQUESTED** [review]
I forgot to mention, can you update the syntax in your PR description please.

**Comment by mgehre-amd** [comment]
> I forgot to mention, can you update the syntax in your PR description please.

Good catch, done!

**Review by simon-camp — APPROVED** [review]
Awesome, thanks!

**Pull request merged**